src/bin/inx-chronicle/api/stardust/explorer/routes.rs

```diff
@@ -164,3 +166,38 @@ async fn block_children(
         children,
     })
 }
+
+async fn milestones(
+    database: Extension<MongoDb>,
+    MilestonesPagination {
+        start_timestamp,
+        end_timestamp,
+        sort,
+        page_size,
+        cursor,
+    }: MilestonesPagination,
+) -> ApiResult<MilestonesResponse> {
+    let mut record_stream = database
+        .collection::<MilestoneCollection>()
+        .get_milestones(start_timestamp, end_timestamp, sort, page_size + 1, cursor)
+        .await?;
+
+    // Take all of the requested records first
+    let items = record_stream
+        .by_ref()
+        .take(page_size)
+        .map_ok(Into::into)
+        .try_collect()
+        .await?;
+
+    // If any record is left, use it to make the paging state
+    let cursor = record_stream.try_next().await?.map(|rec| {
+        MilestonesCursor {
+            milestone_index: rec.index,
+            page_size,
+        }
+        .to_string()
+    });
+
+    Ok(MilestonesResponse { items, cursor })
```

Comment:
That might be wrong! What happens with the successfully collected `items` if `try_next` on the `record_stream` fails (i.e. no record is left) ?